src/software/backend/input/network/networking/proto_multicast_listener.h

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <boost/asio.hpp>
+#include <boost/bind.hpp>
+#include <string>
+
+#include "software/proto/messages_robocup_ssl_wrapper.pb.h"
+#include "software/proto/ssl_referee.pb.h"
+
+template <class ReceiveProto>
+class ProtoMulticastListener
+{
+   public:
+    /**
+     * Creates an ProtoMulticastListener that will listen for ReceiveProto packets from
+     * the network on the given address and port. For every ReceiveProto packet received,
+     * the receive_callback will be called to perform any operations desired by the caller
+     *
+     * @param io_service The io_service to use to service incoming ReceiveProto data
+     * @param ip_address The ip address of the multicast group on which to listen for
+     * the given ReceiveProto packets
+     * @param port The port on which to listen for ReceiveProto packets
+     * @param receive_callback The function to run for every ReceiveProto packet received
+     * from the network
+     */
+    ProtoMulticastListener(boost::asio::io_service& io_service, std::string ip_address,
+                           unsigned short port,
+                           std::function<void(ReceiveProto)> receive_callback);
+
+   private:
+    /**
+     * The function that is called to process any data received by the io_service. This
+     * function gets automatically called by the io_service for EVERY packet received by
+     * the network. These functions are handled synchronously, so we DO NOT need to worry
+     * about concurrency or thread-safety in this function. Because this function also
+     * calls the provided receive_callback, this means the receive_callback also does not
+     * need to be thread-safe
+     *
+     * @param error The error code obtained when receiving the incoming data
+     * @param num_bytes_received How many bytes of data were received
+     */
+    void handleDataReception(const boost::system::error_code& error,
+                             size_t num_bytes_received);
+
+    // A UDP socket that we listen on for ReceiveProto messages from the network
+    boost::asio::ip::udp::socket socket_;
+    // The endpoint for the sender
+    boost::asio::ip::udp::endpoint sender_endpoint_;
+
+    // The maximum length of the buffer we use to receive data packets from the network.
+    // 9000 bytes is the maximum transfer unit (MTU) size we can set on our interfaces.
```

Comment:
Can we specify where / what these limiting interfaces are? @akhilveeraghanta mentioned from firmware but can we be more specific and add more detail here? I'm nitpicking so that people in the future can really go look up the limiting interfaces / find them directly.

Reply:
Anything thats over 1500 bytes of payload is a jumbo frame. Usually its 9000 bytes for CPU cycles/performance reasons

From wikipedia: 

The use of 9000 bytes as preferred payload size for jumbo frames arose from discussions within the Joint Engineering Team of Internet2 and the U.S. federal government networks.[7] Their recommendation has been adopted by all other national research and education networks. In order to meet this mandatory purchasing criterion, manufacturers have in turn adopted 9000 bytes as the conventional MTU size, with a jumbo frame size of at least 9018/9022 bytes (without/with IEEE 802.1Q field).[citation needed] Most Ethernet equipment can support jumbo frames up to 9216 bytes.[8]

https://en.wikipedia.org/wiki/Jumbo_frame

Reply:
So we can easily go over 9000, with a slight hit on the CPU cycles because it would need to reassemble the fragments, which requires no extra code for us to support.

Reply:
Yea since we receive into the buffer and then immediately decode the proto, if the buffer is too small we'd likely fail to decode. Not sure if this throws an error or not. I also wonder if the `num_bytes_received` can report a number of bytes larger than the buffer. It may be good to see if that's possible, and add our own error handling / logging to we can be made aware if the buffer is too small for whatever reason in the future (ie. we get a big message).

Reply:
Sorry I'm really sticking on this but I'd really like this comment to be more specific. Specify that it's more a limitation of our firmware interfaces, and link to the jumbo frame page at the very least.

Reply:
yeah I agree that it's important. I've added warning: from testing with a small buffer size, num_bytes_received will report the max_buffer_length if the message size is too big (sadly no error code though). The warning will be triggered if num_bytes_received == max_buffer_length

Reply:
Ok hopefully final comment :tm: here. Given all the discussion we've had, I think I understand what's going on now so here's my suggestion for the comment here:

```
// The maximum length of the buffer we use to receive data packets from the network.
// 9000 bytes is the maximum transfer unit (MTU) size we can set on our firmware interfaces,
// and is also the maximum size of a jumbo frame before the frame will be fragmented when it is sent.
// https://en.wikipedia.org/wiki/Jumbo_frame
// We make the assumption that any messages we define, as well as the proto messages used
// by the SSL, will be less than this value.
```

Trying to make the assumptions more clear, more clear what interfaces are the limitation, and the _we_ aren't doing the fragmenting.

Reply:
done